File: src/main/java/org/moinex/entities/CreditCardCredit.java
/*
 * Filename: CreditCardCredit.java
 * Created on: March  5, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.entities;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;
import org.moinex.util.enums.CreditCardCreditType;

/**
 * Represents a credit card credit
 * A credit card credit is a credit made to a credit card
 */
@Entity
@Table(name = "credit_card_credit")
@Getter
@Setter
@SuperBuilder
@NoArgsConstructor
@AllArgsConstructor
public class CreditCardCredit extends CreditCardTransaction
{
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    @Setter(AccessLevel.NONE)
    private Long id;

    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false)
    private CreditCardCreditType type;
}

File: src/main/kotlin/org/moinex/model/creditcard/CreditCardCredit.kt
package org.moinex.model.creditcard

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import org.moinex.model.enums.CreditCardCreditType
import java.math.BigDecimal
import java.time.LocalDateTime

@Entity
@Table(name = "credit_card_credit")
class CreditCardCredit(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    var id: Int? = null,
    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false)
    var type: CreditCardCreditType,
    creditCard: CreditCard,
    date: LocalDateTime,
    amount: BigDecimal,
    description: String? = null,
) : CreditCardTransaction(
        creditCard = creditCard,
        date = date,
        amount = amount,
        description = description,
    ) {
    override fun toString(): String {
        return "Credit Card Credit [id=$id, type=$type, amount=$amount, date=$date, " +
            "description=$description, creditCardId=${creditCard.id}]"
    }
}
